java/ql/src/Security/CWE/CWE-378/TempDirHijackingVulnerability.ql

```diff
@@ -0,0 +1,90 @@
+/**
+ * @name Temporary Directory Hijacking Vulnerability disclosure
+ * @description Detect temporary directory hijack vulnerability
+ * @kind path-problem
+ * @problem.severity error
+ * @precision very-high
+ * @id java/temp-directory-hijacking
+ */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import DataFlow::PathGraph
+
+/**
+ * All `java.io.File::createTempFile` methods.
+ */
+private class MethodFileCreateTempFile extends Method {
+  MethodFileCreateTempFile() {
+    this.getDeclaringType() instanceof TypeFile and
+    this.hasName("createTempFile")
+  }
+}
+
+private class MethodFileMkdir extends Method {
+  MethodFileMkdir() {
+    getDeclaringType() instanceof TypeFile and
+    hasName("mkdir")
+    or
+    hasName("mkdirs")
+  }
+}
+
+private class MethodFileDelete extends Method {
+  MethodFileDelete() {
+    getDeclaringType() instanceof TypeFile and
+    hasName("delete")
+  }
+}
+
+predicate isDeleteFileExpr(Expr expr) {
+  exists(MethodAccess ma |
+    expr = ma.getQualifier() and
+    ma.getMethod() instanceof MethodFileDelete
+  )
+}
+
+private class TempDirHijackingToDeleteConfig extends DataFlow::Configuration {
+  TempDirHijackingToDeleteConfig() { this = "TempDirHijackingToDeleteConfig" }
+
+  override predicate isSource(DataFlow::Node source) {
+    exists(MethodAccess ma |
+      ma.getMethod() instanceof MethodFileCreateTempFile and
+      source.asExpr() = ma
+    )
+  }
+
+  override predicate isSink(DataFlow::Node sink) { isDeleteFileExpr(sink.asExpr()) }
+}
+
+private class TempDirHijackingFromDeleteConfig extends DataFlow2::Configuration {
+  TempDirHijackingFromDeleteConfig() { this = "TempDirHijackingFromDeleteConfig" }
+
+  override predicate isSource(DataFlow::Node source) { isDeleteFileExpr(source.asExpr()) }
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(MethodAccess ma |
+      ma.getMethod() instanceof MethodFileMkdir and
+      ma.getQualifier() = sink.asExpr()
+    )
```

Comment:
`Files.createDirectories` should also be considered here.

Reply:
That's actually not true, both are unsafe.

Reply:
I decided not to do this, because this would be taint flow, instead of data. flow. I don't think the added overhead of taint flow is worth adding these two methods that are probably not commonly leveraged